plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/index/HudiPartitionStatsIndexSupport.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hudi.query.index;
+
+import io.airlift.log.Logger;
+import io.trino.plugin.hudi.util.TupleDomainUtils;
+import io.trino.spi.predicate.TupleDomain;
+import org.apache.hudi.avro.model.HoodieMetadataColumnStats;
+import org.apache.hudi.common.model.FileSlice;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.hash.ColumnIndexID;
+import org.apache.hudi.metadata.HoodieTableMetadata;
+import org.apache.hudi.metadata.HoodieTableMetadataUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static io.trino.plugin.hudi.util.TupleDomainUtils.hasSimpleNullCheck;
+
+public class HudiPartitionStatsIndexSupport
+        extends HudiColumnStatsIndexSupport
+{
+    private static final Logger log = Logger.get(HudiColumnStatsIndexSupport.class);
+
+    public HudiPartitionStatsIndexSupport(HoodieTableMetaClient metaClient)
+    {
+        super(log, metaClient);
+    }
+
+    @Override
+    public Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(HoodieTableMetadata metadataTable, Map<String, List<FileSlice>> inputFileSlices, TupleDomain<String> regularColumnPredicates)
+    {
+        return Map.of();
+    }
+
+    public Optional<List<String>> prunePartitions(HoodieTableMetadata metadataTable, TupleDomain<String> regularColumnPredicates)
+    {
+        // Filter out predicates containing simple null checks (`IS NULL` or `IS NOT NULL`)
+        TupleDomain<String> filteredRegularPredicates = regularColumnPredicates.filter((_, domain) -> !hasSimpleNullCheck(domain));
+
+        // Sanity check, if no regular domains, return immediately
+        if (filteredRegularPredicates.getDomains().isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<String> regularColumns = new ArrayList<>(filteredRegularPredicates.getDomains().get().keySet());
+
+        // Get columns to filter on
+        List<String> encodedTargetColumnNames = regularColumns
+                .stream()
+                .map(col -> new ColumnIndexID(col).asBase64EncodedString()).toList();
+
+        // Map of partition stats keyed by partition name
+        // If some columns in encodedTargetColumnNames is not available in partition stats index, partition will not be pruned
```

Comment:
If i understand the below stream operation correctly, by collecting only partitions with stats and then returning that list, any partition without stats will be dropped entirely. Should we:
1. Fetch the full set of partitions from the metadata table.
2. For each partition, check that stats exist for _all_ requested columns.
3. Only evaluate the predicate on those "complete" partitions; include all others in the result (i.e. don’t prune them).

Reply:
Ahh, good catch! I will add this in tomorrow!.